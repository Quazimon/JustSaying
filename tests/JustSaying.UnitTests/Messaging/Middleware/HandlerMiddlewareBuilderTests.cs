using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using JustSaying.Fluent;
using JustSaying.Messaging.Middleware.Handle;
using JustSaying.TestingFramework;
using JustSaying.UnitTests.Messaging.Channels.Fakes;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Shouldly;
using Xunit;
using Xunit.Abstractions;

namespace JustSaying.UnitTests.Messaging.Middleware
{
    public class HandlerMiddlewareBuilderTests
    {
        private readonly FakeServiceResolver _resolver;

        public HandlerMiddlewareBuilderTests()
        {
            _resolver = new FakeServiceResolver();
        }

        [Fact]
        public async Task ThreeMiddlewares_ShouldExecuteInCorrectOrder()
        {
            var callRecord = new List<string>();

            void Before(string id) => callRecord.Add($"Before_{id}");
            void After(string id) => callRecord.Add($"After_{id}");

            var outer = new TrackingMiddleware("outer", Before, After);
            var middle = new TrackingMiddleware("middle", Before, After);
            var inner = new TrackingMiddleware("inner", Before, After);

            var middleware = new HandlerMiddlewareBuilder(_resolver, _resolver)
                .Configure(pipe =>
                {
                    pipe.Use(outer);
                    pipe.Use(middle);
                    pipe.Use(inner);
                }).Build();

            var context = new HandleMessageContext(new SimpleMessage(),
                typeof(SimpleMessage),
                "a-fake-queue");

            await middleware.RunAsync(context,
                ct =>
                {
                    callRecord.Add("HandledMessage");
                    return Task.FromResult(true);
                },
                CancellationToken.None);

            var record = string.Join(Environment.NewLine, callRecord);

            record.ShouldMatchApproved(c => c.SubFolder("Approvals"));
        }
    }
}
