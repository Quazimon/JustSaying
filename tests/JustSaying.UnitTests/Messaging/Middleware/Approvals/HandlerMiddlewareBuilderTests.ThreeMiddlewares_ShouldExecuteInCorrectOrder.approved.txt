Before_outer
Before_middle
Before_inner
HandledMessage
After_inner
After_middle
After_outer